File: CombinationSum39.java
import java.util.LinkedList;
import java.util.List;

class CombinationSum39 {


    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        //solution
        List<List<Integer>> result = new LinkedList<>();

        //state
        List<Integer> state = new LinkedList<>();

        //backtrackig
        backTrack(result,state,candidates,target,0,0);
        return result;
    }

    private void backTrack(List<List<Integer>> result, List<Integer> state, int[] candidates, int target, int index,
            int sum) {
    
                //isvalid
                if(sum>target) return;

                //isgoal
                if(sum==target) result.add(new LinkedList<>(state));

                //combinations
                for(int i=index; i < candidates.length; i++){
                    state.add(candidates[i]);
                    backTrack(result, state, candidates, target, i++, sum + candidates[i]);
                    state.remove(state.size()-1);
                
            }
            }


public static void main(String[] args) {
    
    CombinationSum39 cs = new CombinationSum39();
    int arr[] = {2,3,6,7};    
    for(List<Integer> l : cs.combinationSum(arr, 7)){
        System.out.println();
        for(int i : l){
            System.out.print(i);
        }
    }
}

}

File: CombinationSum-39.py
# Time Complexity : O(n^n)
# Space Complexity : O(n)  element added to curr
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

#add a candidate and check for target if not remove it
#[2,2,2,2] -> [2,2,2,3] -> [2,2,2,6] -> [2,2,2] (remove)
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        def backtrack(curr, idx ,target):
            if target == 0:
                res.append(list(curr))
            elif target < 0:
                return
            for i in range(idx, len(candidates)):
                curr.append(candidates[i])
                backtrack(curr, i, target - candidates[i])
                curr.remove(candidates[i])

        backtrack([], 0, target)
        return res
